sdks/python/gen_protos.py

```diff
@@ -15,40 +15,100 @@
+"""
+Generates Python proto modules and grpc stubs for Beam protos.
+"""
+
 import contextlib
 import glob
 import inspect
 import logging
-import multiprocessing
 import os
 import platform
 import re
 import shutil
 import subprocess
 import sys
 import time
-import warnings
+from collections import defaultdict
 from importlib import import_module
 
 import pkg_resources
 
+LOG = logging.getLogger()
+LOG.setLevel(logging.INFO)
+
+LICENSE_HEADER = """
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+
+NO_PROMISES_NOTICE = """
+\"\"\"
+For internal use only; no backwards-compatibility guarantees.
+Automatically generated when running setup.py sdist or build[_py].
+\"\"\"
+"""
+
+
+def clean_path(path):
+  return os.path.realpath(os.path.abspath(path))
+
+
+# These paths are relative to the project root
 BEAM_PROTO_PATHS = [
-    os.path.join('..', '..', 'model', 'pipeline', 'src', 'main', 'proto'),
-    os.path.join('..', '..', 'model', 'job-management', 'src', 'main', 'proto'),
-    os.path.join('..', '..', 'model', 'fn-execution', 'src', 'main', 'proto'),
-    os.path.join('..', '..', 'model', 'interactive', 'src', 'main', 'proto'),
+    os.path.join('model', 'pipeline', 'src', 'main', 'proto'),
+    os.path.join('model', 'job-management', 'src', 'main', 'proto'),
+    os.path.join('model', 'fn-execution', 'src', 'main', 'proto'),
+    os.path.join('model', 'interactive', 'src', 'main', 'proto'),
 ]
 
-PYTHON_OUTPUT_PATH = os.path.join('apache_beam', 'portability', 'api')
+PYTHON_SDK_ROOT = os.path.dirname(clean_path(__file__))
+PROJECT_ROOT = clean_path(os.path.join(PYTHON_SDK_ROOT, '..', '..'))
+PYTHON_OUTPUT_PATH = os.path.join(
+    PYTHON_SDK_ROOT, 'apache_beam', 'portability', 'api')
 
 MODEL_RESOURCES = [
-    os.path.normpath('../../model/fn-execution/src/main/resources'\
-            + '/org/apache/beam/model/fnexecution/v1/standard_coders.yaml'),
+    os.path.normpath((
+        'model/fn-execution/src/main/resources/org/'
+        'apache/beam/model/fnexecution/v1/standard_coders.yaml')),
 ]
 
 
-def generate_urn_files(log, out_dir):
+class PythonPath(object):
+  def __init__(self, path: str, front: bool = False):
+    self._sys_path = sys.path.copy()
```

Comment:
Technically, this should probably be initialized on at `__enter__` rather that here.

Reply:
fixed